### PR TITLE
2.1 Sample Updates (Round 11)

### LINKED ISSUE #6398: The sample is misleading
Hi
If you try the above example and start the published app from the commandline you will end up with this error.
Unhandled Exception: System.FormatException: Value for switch '--console' is missing
I found this on Stack Overflow https://stackoverflow.com/questions/48992554/net-core-app-command-line-format-exception?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa
It solved the problem.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b0bd8cb8-7e28-8c4d-a550-3aea8f3ab4dd
* Version Independent ID: 4b60d24e-ef05-15e5-10b4-483094081bf3
* Content: [Host ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.1&tabs=aspnetcore2x)
* Content Source: [aspnetcore/host-and-deploy/windows-service.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/windows-service.md)
* Service: **aspnet-core**
* GitHub Login: @tdykstra
* Microsoft Alias: **tdykstra**

### LINKED ISSUE #6876: Hosting as a windows service no longer requires .NET Framework
In the "Prerequisites" section, the first sentence is: "The app must run on the .NET Framework runtime." This seems to no longer be the case. I used this guide to host a .NET Core 2.0/ASP.NET Core 2.0 app as a Windows service.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b0bd8cb8-7e28-8c4d-a550-3aea8f3ab4dd
* Version Independent ID: 4b60d24e-ef05-15e5-10b4-483094081bf3
* Content: [Host ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.1&tabs=aspnetcore2x)
* Content Source: [aspnetcore/host-and-deploy/windows-service.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/windows-service.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **tdykstra**

### PR DESCRIPTION
**Pull request by guardrex**
Addresses #5495 
Fixes #6398 
Fixes #6876

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-1.0&branch=pr-en-us-6860)

* Windows Service topic updates
* Sample to 2.1 and Razor Pages

@Tratcher AppInsights wasn't necessary, so I removed it. However, I wonder if that was there to gauge the use of Core for Windows Services. Do you know why it was in the original sample?

Thanks to @notes2c and @waxingsatirical for suggesting improvements to the topic.

### CODE REVIEW

**Comment by Tratcher** [comment]
I don't know why it was included before. It was likely copied from somewhere else.

**Comment by guardrex** [comment]
@Tratcher There's a report by @gregbair that a Windows Service app can target `netcoreapp2.1` ... I didn't see it in a blog post or the release notes. Can u confirm it? ... I'll patch this PR and the sample before you review this.

**Comment by Tratcher** [comment]
It's true. See https://docs.microsoft.com/en-us/dotnet/core/porting/windows-compat-pack and https://www.nuget.org/packages/Microsoft.AspNetCore.Hosting.WindowsServices/

**Comment by guardrex** [comment]
Ok, thanks. I'll take care of it.

**Comment by guardrex** [comment]
@Tratcher In addition to updating to `netcoreapp2.1`, I fleshed out the instruction for creating, starting, querying, stopping, and uninstalling the service.

It's ready whenever you are. No rush on my end ... I'm on ... what do they call it ... *vacation???* ... sounds kind'a lame to me, but I guess I'll give it a try ... *maybe*. :grinning:

File: aspnetcore/host-and-deploy/windows-service.md [hunk 1]
```diff
@@ -5,110 +5,175 @@ description: Learn how to host an ASP.NET Core app in a Windows Service.
...
 
   [!code-xml[](windows-service/sample/AspNetCoreService.csproj?range=3-6)]
 
-  When creating a project in Visual Studio, use the **ASP.NET Core Application (.NET Framework)** template.
+  To create a project in Visual Studio suitable for modification into a Windows Service app, use the **ASP.NET Core Application** template:
+    1. Select **File** > **New** > **Project**.
+    1. Select the **ASP.NET Core Web Application** template. Name the app and select **OK**.
+    1. In the first drop-down, select **.NET Framework**. Leave **Web Application** selected. Select **OK**.
 
 * If the app receives requests from the Internet (not just from an internal network), it must use the [HTTP.sys](xref:fundamentals/servers/httpsys) web server (formerly known as [WebListener](xref:fundamentals/servers/weblistener) for ASP.NET Core 1.x apps) rather than [Kestrel](xref:fundamentals/servers/kestrel). Use of IIS in a reverse proxy server configuration with Kestrel is an option for edge deployments. For more information, see [When to use Kestrel with a reverse proxy](xref:fundamentals/servers/kestrel#when-to-use-kestrel-with-a-reverse-proxy).
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/windows-service.md]: This non-edge kestrel restriction is outdated.

File: aspnetcore/host-and-deploy/windows-service.md [hunk 2]
```diff
@@ -5,110 +5,175 @@ description: Learn how to host an ASP.NET Core app in a Windows Service.
...
+1. Publish the app to a folder. Use [dotnet publish](/dotnet/articles/core/tools/dotnet-publish) or a [Visual Studio publish profile](xref:host-and-deploy/visual-studio-publish-profiles) that publishes to a folder.
 
-   Open a command shell with administrative privileges to use the [sc.exe](https://technet.microsoft.com/library/bb490995) command-line tool to create and start a service. If the service is named MyService, published to `c:\svc`, and named AspNetCoreService, the commands are:
+   To publish the sample app from the command line, run the following command in a console window from the project folder:
+
+   ```console
+   dotnet publish --configuration Release --output c:\\svc
+   ```
+
+   The double-backslash escapes a backslash in the command-line argument for the output path. The project is published to the *c:\\svc* folder.
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/windows-service.md]: Where is this needed? I don't need double slashes on cmd.

File: aspnetcore/host-and-deploy/windows-service.md (hunk 1 above)
Comment (guardrex) [reply aspnetcore/host-and-deploy/windows-service.md]: Cool beans ... Your prior two points eliminate all prereqs.

File: aspnetcore/host-and-deploy/windows-service.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/host-and-deploy/windows-service.md]: I must have cut-and-paste from Line 59!

*I really **DO** need a vacation!* 🏖

**Comment by guardrex** [comment]
@Tratcher Ready for another :eye:.

**Comment by Tratcher** [comment]
So you fixed #6398 by no longer passing args to the WebHostBuilder?

**Comment by guardrex** [comment]
@Tratcher I did it this way ... (you can see it in the diff for *Program.cs*) ...

```csharp
var webHostArgs = args.Where(arg => arg != "--console").ToArray();

var host = WebHost.CreateDefaultBuilder(webHostArgs)
    ...
```

**Comment by guardrex** [comment]
:point_up: Stolen from our friend @dasMulli.

**Comment by Tratcher** [comment]
Nevermind, I was looking at the wrong TFM

File: aspnetcore/host-and-deploy/windows-service.md [hunk 3]
```diff
@@ -5,110 +5,160 @@ description: Learn how to host an ASP.NET Core app in a Windows Service.
...
+ms.date: 06/04/2018
 ms.prod: aspnet-core
 ms.technology: aspnet
 ms.topic: article
 uid: host-and-deploy/windows-service
 ---
 # Host ASP.NET Core in a Windows Service
 
-By [Tom Dykstra](https://github.com/tdykstra)
+By [Tom Dykstra](https://github.com/tdykstra) and [Luke Latham](https://github.com/guardrex)
```
Comment (Rick-Anderson) [inline aspnetcore/host-and-deploy/windows-service.md]: Put Luke first.

File: aspnetcore/host-and-deploy/windows-service.md [hunk 4]
```diff
@@ -5,110 +5,160 @@ description: Learn how to host an ASP.NET Core app in a Windows Service.
...
-   # [ASP.NET Core 2.x](#tab/aspnetcore2x/)
+   To publish the sample app from the command line, run the following command in a console window from the project folder:
 
-   [!code-csharp[](windows-service/sample/Program.cs?name=ServiceOnly&highlight=3-4,7,12)]
+   ```console
+   dotnet publish --configuration Release --output c:\svc
+   ```
 
-   # [ASP.NET Core 1.x](#tab/aspnetcore1x/)
+1. Open a command shell with administrative privileges to use the [sc.exe](https://technet.microsoft.com/library/bb490995) command-line tool to create the service. Use the `sc create <SERVICE_NAME> binPath= "<PATH_TO_SERVICE_EXECUTABLE>"` command. The `binPath` value is the path to the app's executable, which includes the executable file name. **The space between the equal sign and the quote character that starts the path is required.**
```
Comment (Rick-Anderson) [inline aspnetcore/host-and-deploy/windows-service.md]: Open a command shell with administrative privileges. The command shell will .. clean this up - just split the sentence in two. --  to use the [sc.exe](https://technet.microsoft.com/library/bb490995) command-line tool to create the service.

File: aspnetcore/host-and-deploy/windows-service/sample/AspNetCoreService.csproj [hunk 5]
```diff
@@ -1,24 +1,20 @@
...
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.1.1" />
-    <PackageReference Include="Microsoft.AspNetCore" Version="2.0.1" />
-    <PackageReference Include="Microsoft.AspNetCore.Hosting.WindowsServices" Version="2.0.1" />
-    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.0.2" />
-    <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="2.0.1" />
-    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="2.0.0" />
-    <PackageReference Include="Microsoft.VisualStudio.Web.BrowserLink" Version="2.0.1" />
+    <PackageReference Include="Microsoft.AspNetCore" Version="2.1.0" />
```
Comment (Eldorian) [inline aspnetcore/host-and-deploy/windows-service/sample/AspNetCoreService.csproj]: As I was going through this example I noticed I wasn't able to access some of the new 2.1 features like HttpClientFactory on IServiceCollection. I had to change this line of code to include Microsoft.AspNetCore.App to make it happen - which I believe is required if you're migrating from 2.0 to 2.1 (according to the migration doc)

Comment (Eldorian) [inline aspnetcore/host-and-deploy/windows-service/sample/AspNetCoreService.csproj]: As I was going through this example I noticed I wasn't able to access some of the new 2.1 features like HttpClientFactory on IServiceCollection. I had to change this line of code to include Microsoft.AspNetCore.App to make it happen - which I believe is required if you're migrating from 2.0 to 2.1 (according to the migration doc)

Comment (guardrex) [reply aspnetcore/host-and-deploy/windows-service/sample/AspNetCoreService.csproj]: @Eldorian It's not strictly speaking required (in this case, you could probably just add `Microsoft.Extensions.Http`), but it does simplify package management and is what we normally pitch for `netcoreapp2.1` apps. Prior versions of this sample listed packages individually, but I don't see a good reason to stick to that approach.

@Tratcher Any objections on moving to ...

```xml
<PackageReference Include="Microsoft.AspNetCore.App" />
<PackageReference Include="Microsoft.AspNetCore.Hosting.WindowsServices" Version="2.1.0" />
```

Comment (Eldorian) [reply aspnetcore/host-and-deploy/windows-service/sample/AspNetCoreService.csproj]: Ah, thanks for that explanation on what I could have done to make it work - I'm pretty new to ASP.NET Core in general and got into it shortly before 2.1 released, so everything I've been reading in the documents to migrate from 2.0 to 2.1 mention Microsoft.AspNetCore.App that in the package reference. 

Just makes more sense to me then to have it be consistent as I'm going through different documentation :)

Comment (guardrex) [reply aspnetcore/host-and-deploy/windows-service/sample/AspNetCoreService.csproj]: The basics on .NET Standard and package management are over in the .NET Core docs. That doc set kicks off over at: https://docs.microsoft.com/dotnet/core/

**Pull request merged**